### PR TITLE
Limiting autonat service responses/startup

### LINKED ISSUE #32: Only dial back to the source IP
We should consider only dialing back to the source IP address to avoid DoS attacks.

Unfortunately, this won't allow us to test IPv6 when the original connection is IPv4 and vice versa.

### LINKED ISSUE #36: Rate Limiting
Ideally, we'd be able to enable AutoNAT on all nodes. However, to do this, we need to:

1. Fix #32. 
2. Severely rate-limit requests, telling most nodes to go away. This should be configurable.

### LINKED ISSUE #38: Clear addresses before dialing back
Otherwise, we'll try old addresses in addition to the new ones.

### LINKED ISSUE #39: Limit the number of addresses we'll try dialing
We should limit this to, say, 16 at the most.

### LINKED ISSUE #43: Avoid turning on until we learn our NAT status
This should help prevent NATed nodes on the same network from confirming each other. This isn't _perfect_ as one node could have configured port forwarding but it should help.

We _should_ be able to do this by listening for an autonat event.

### PR DESCRIPTION
**Pull request by willscott**
* limits addresses for a peer (at most 16 accepted) - fix #39 
* clear addresses before dialing back - fix #38
* global rate limit of 30 responses per (1 - 1.25 min) - fix #36
* only dial back on the source IP - fix #32 
* Delay startup until reachability public - fix #43

### CODE REVIEW

File: svc.go [hunk 1]
```diff
@@ -51,9 +52,21 @@ func NewAutoNATService(ctx context.Context, h host.Host, opts ...libp2p.Option)
 		dialer: dialer,
 		reqs:   make(map[peer.ID]int),
 	}
-	h.SetStreamHandler(autonat.AutoNATProto, as.handleStream)
 
-	go as.resetRateLimiter()
+	s, err := h.EventBus().Subscribe(&event.EvtLocalRoutabilityPublic{})
```
Comment (aarshkshah1992) [inline svc.go]: @willscott @Stebalien 

**1.** Does this mean that every AutoNat server will also have to be an AutoNat client (ONLY the client has the capability to emit this event for now) ? If yes, who "starts" the client on the server ?

**2.** From the linked issue:
```
This isn't perfect as one node could have configured port forwarding 
but it should help.
```
How can port forwarding spoil the party here ?

**3.** As @willscott  has already mentioned, we should have a functional Option(false by default) to set this to true if we are already confident during initialization that we have a public address.

Comment (willscott) [reply svc.go]: > every AutoNat server will also have to be an AutoNat client

Yes. If we intend that nodes offering AutoNat Service are publicly reachable and valid DHT participants, then they need to learn that.

> why is port forwarding problematic?

I think the scenarios with port forwarding we want to understand are:
* I tell my router "public 4001 -> local 4001", and then configure IPFS locally to be on port 4001. Outbound connections, even if sourced from, say, `192.168.0.1:4001`, will get NATed by my router to `publicIP:<ephemeralport>`. We should potentially dial back on `publicIP:<low ports seen with private IPs>` to identify these cases.

Comment (Stebalien) [reply svc.go]: >  If yes, who "starts" the client on the server?

I believe we'll just have to start it on all nodes a the libp2p layer. But this does raise an interesting issue with the events system: dependencies.

> I think the scenarios with port forwarding we want to understand are

Note: If we fix https://github.com/libp2p/go-libp2p-autonat/issues/47 and https://github.com/libp2p/go-libp2p-autonat-svc/issues/44 (which we can do by listening for NAT routability events), this should be a non-issue.

File: svc.go [hunk 2]
```diff
@@ -24,36 +29,49 @@ const P_CIRCUIT = 290
...
+		h:            h,
+		dialer:       dialer,
+		globalReqMax: AutoNATGlobalThrottle,
+		reqs:         make(map[peer.ID]int),
 	}
-	h.SetStreamHandler(autonat.AutoNATProto, as.handleStream)
 
-	go as.resetRateLimiter()
+	if forceEnabled {
+		as.globalReqMax = 0
```
Comment (aarshkshah1992) [inline svc.go:79]: Why should `globalReqMax` be 0 if forceEnabled is true ? We might be sure that we have public Reachability but would still want the global throttling.

File: svc.go [hunk 3]
```diff
@@ -24,36 +29,49 @@ const P_CIRCUIT = 290
 var (
 	AutoNATServiceDialTimeout   = 15 * time.Second
 	AutoNATServiceResetInterval = 1 * time.Minute
+	AutoNATServiceResetJitter   = 15 * time.Second
 
-	AutoNATServiceThrottle = 3
+	AutoNATServiceThrottle  = 3
+	AutoNATGlobalThrottle   = 30
+	AutoNATMaxPeerAddresses = 16
```
Comment (aarshkshah1992) [inline svc.go:46]: Documentation for all exported types. On a related note, why should we even export these types ?

File: svc.go [hunk 4]
```diff
@@ -162,24 +210,35 @@ func (as *AutoNATService) skipDial(addr ma.Multiaddr) bool {
 		return true
 	}
 
+	// Skip dialing addresses we believe are the local node's
```
Comment (aarshkshah1992) [inline svc.go:203]: Is this to prevent an AutoNAT client from getting a false positive if the Server is on the same network ?

File: svc.go [hunk 5]
```diff
@@ -162,24 +210,35 @@ func (as *AutoNATService) skipDial(addr ma.Multiaddr) bool {
...
+
 	return false
 }
 
 func (as *AutoNATService) doDial(pi peer.AddrInfo) *pb.Message_DialResponse {
 	// rate limit check
 	as.mx.Lock()
 	count := as.reqs[pi.ID]
-	if count >= AutoNATServiceThrottle {
+	if count >= AutoNATServiceThrottle || (as.globalReqMax > 0 && as.globalReqs >= as.globalReqMax) {
```
Comment (aarshkshah1992) [inline svc.go:217]: Can we have a unit test for the global throttling ?

**Review by aarshkshah1992 — CHANGES_REQUESTED** [review]

File: svc.go [hunk 6]
```diff
@@ -188,28 +247,52 @@ func (as *AutoNATService) doDial(pi peer.AddrInfo) *pb.Message_DialResponse {
...
-		return newDialResponseError(pb.Message_E_INTERNAL_ERROR, "internal service error")
-	}
-
-	ra := conns[0].RemoteMultiaddr()
+	ra := conn.RemoteMultiaddr()
 	as.dialer.Network().ClosePeer(pi.ID)
 	return newDialResponseOK(ra)
 }
 
+func (as *AutoNATService) enableWhenPublic() {
```
Comment (aarshkshah1992) [inline svc.go:245]: Please can you add a unit test that emits mock Reachability events on the bus & ensures this function works as expected ? We should test both the  `SetStreamHandler ` &  `RemoveStreamHandler` branches.

**Review by aarshkshah1992 — CHANGES_REQUESTED** [review]

File: svc.go (hunk 2 above)
Comment (willscott) [reply svc.go:79]: there are a lot of potential knobs. I felt like the most likely thing - the aggressive mode that would be wanted on a bootstrap node - was to ensure that the autonat service was on and responding positively to queries. if you force enable it, your default expectation is probably that it's going to be responding to all inbound requests.

File: svc.go (hunk 4 above)
Comment (willscott) [reply svc.go:203]: yep

**Review by aarshkshah1992 — APPROVED** [review]

**Comment by aarshkshah1992** [comment]
LGTM.

File: svc.go [hunk 7]
```diff
@@ -96,6 +125,26 @@ func (as *AutoNATService) handleStream(s network.Stream) {
 	}
 }
 
+// Optimistically extract the net.IP host from a multiaddress.
+// TODO: use upstream manet.ToIP
```
Comment (Stebalien) [inline svc.go]: released now.

**Comment by Stebalien** [comment]
I'd update the deps and merge this.

**Pull request merged**